### PR TITLE
ARROW-10578: [C++] Comparison kernels crashing for string array with null string scalar

### PR DESCRIPTION
**Pull request by KirillLykov**

### CODE REVIEW

**Comment by kou** [comment]
@github-actions autotune

File: python/pyarrow/tests/test_compute.py [hunk 1]
```diff
@@ -803,12 +836,17 @@ def con(values): return pa.array(values)
...
-    scalar = pa.array([2]).sum()
+    scalar = pa.scalar(2)
 
     result = pc.equal(arr, scalar)
     assert result.equals(con([False, True, False, None]))
 
+    if typ == "array":
+        nascalar = pa.scalar(None, type="int64")
+        result = pc.equal(arr, nascalar)
+        isnull = pc.is_null(result)
```
Comment (pitrou) [inline python/pyarrow/tests/test_compute.py:847]: Or simply: `assert result.to_pylist() == [None, None, None, None]`

**Review by pitrou — COMMENTED** [review]
Thank you for this PR! See comment below; also, can you fix Python style issues noticed in CI (see failed lint job)?

File: python/pyarrow/tests/test_compute.py (hunk 1 above)
Comment (KirillLykov) [reply python/pyarrow/tests/test_compute.py]: All done

**Review by jorisvandenbossche — APPROVED** [review]
Thanks!

**Pull request closed without merging**